File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    visited = [[False for _ in range(101)] for _ in range(101)]
    check = [[0 for _ in range(101)] for _ in range(101)]
    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]
    
    def bfs(x, y):
        q = deque()
        q.append([x,y])
        visited[x][y] = True
        check[x][y] += 1
        while q:
            x, y = q.popleft()
            
            for i in range(4):
                nextX = dx[i] + x
                nextY = dy[i] + y
                
                if 0 <= nextX < len(maps) and 0<= nextY < len(maps[0]) and maps[nextX][nextY] != 0 and not visited[nextX][nextY]:
                    q.append([nextX, nextY])
                    visited[nextX][nextY] = True
                    check[nextX][nextY] = check[x][y] + 1
    bfs(0,0)
    print(check)
    if not visited[len(maps)-1][len(maps[0])-1]:
        return -1
    else:
        return check[len(maps)-1][len(maps[0])-1]

    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static int[] dx = {1, 0, -1, 0};
	static int[] dy = {0, 1, 0, -1};
    
    public int solution(int[][] maps) {
        
        int[] dx = {1, 0, -1, 0};
        int[] dy = {0, 1, 0, -1};
        
        return bfs(maps);
    }
    
    static int bfs(int[][] maps) {
		Queue<int[]> q = new LinkedList<>();

		q.offer(new int[] {0, 0, 1});

		while(!q.isEmpty()) {
			int[] data = q.poll();
			int y = data[0];
			int x = data[1];
			int count = data[2];

			if (y == maps.length -1 && x == maps[0].length -1) { // 상대 진영에 도착하면 바로 return
				return count;
			}

			for (int i = 0; i < 4; i++) {
				int nextY = y + dy[i];
				int nextX = x + dx[i];

				if (0 <= nextY && nextY < maps.length && 0 <= nextX && nextX < maps[0].length && maps[nextY][nextX] == 1) {
					maps[nextY][nextX] = 0;
					q.offer(new int[] {nextY, nextX, count + 1});
				}
			}
		}

		return -1;
	}
}